### ISSUE TITLE
clcache binary choice can be fuzzy

### ISSUE THREAD

**Issue by deajan** [Nuitka/Nuitka#906]
Hello,

I've found a "maybe can be improved" situation in Nuitka.
When compiling, nuitka will select *any* clcache binary it finds, even if it doesn't belong to the current python we're running.

In my case,  while I compiled with the interpreter found in `C:\Python38-64`, nuitka relied on clcache found in the interpreter `C:\Python37-32` since the latter one had clcache installed.
I use to set `NUITKA_CLCACHE_BINARY=C:\Python38-64\Scripts\clcache.exe` in my build scripts when clcache is detected in current interpreter, but since it wasn't, that variable wasn't set.

While I did read the logic in ` nuitka/build/SconsCaching.py` which tries to find clcache belonging to the current interpreter first,  it might be useful to emit a warning message when the clcache binary belongs to another interpreter than the current one, just to make sure to be informed.

I'd be happy to make the corresponding PR if you think this makes any sense.

[SIDEQUEST]
Reasons for this remark were some clcache permission errors I encountered (see below).
While the error does not come from Nuitka itself (see https://github.com/frerich/clcache/pull/334/files), the fact that nuitka did use another interpreter's clcache made me jump to conclusions too fast in the first place.

It seems that clcache development has stalled, and while the fix described above is 2 years old, there is no newer clcache package on PyPI (github repo has been archived).
I've checked the forks, but none seems to be 'official'.

Just some ideas... How about shipping nuitka with the latest clcache code, or maybe I can create a 'clcache-nuitka' PyPI package that includes last fixes ?
[/SIDEQUEST]

Best regards.

* Nuitka version, full Python version and Platform (Windows, OSX, Linux ...)

```
0.6.10.1
Python: 3.8.6 (tags/v3.8.6:db45529, Sep 23 2020, 15:52:53) [MSC v.1927 64 bit (AMD64)]
Executable: c:\Python38-64\python.exe
OS: Windows
Arch: x86_64
```
* How did you install Nuitka and Python (pip, anaconda, deb, rpm, from source,
  what is a virtualenv ...), this is very important usually.

`c:\Python38-64\python.exe -m pip install nuitka`

* Many times when you get an error from Nuitka, your setup may be so special
  that even a "hello world" program will not work, please try that and report
  it instead.

* If possible please supply a [Short, Self Contained, Correct, Example](http://sscce.org/)
  that demonstrates the issue i.e a small piece of code which reproduces
  the issue and can be run with out any other (or as few as possible)
  external dependencies.

Not needed here I guess.

Here's my test scenario:

Compilation command line
```
"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat" x64 && C:\Python38-64\python.exe -m nuitka --python-flag=-OO --assume-yes-for-downloads --standalone --plugin-enable=multiprocessing --include-package=urllib3 --windows-product-version="1.2.3.4" --windows-file-description="My Supertool" --windows-company-name="Company" --windows-icon-from-ico="C:\GIT\PROJECT\SSRC\NSIS\some.ico" --output-dir="C:\GIT\PROJECT\SSRC\PYTHON_BUILD64" C:\GIT\PROJECT\PythonSRC\mysupertool.py
```

Nuitka will output the following

```
Nuitka:INFO: Starting Python compilation.
Nuitka-Plugins:INFO: multiprocessing: Injecting pre-module load code for module 'multiprocessing':
Nuitka-Plugins:INFO: multiprocessing:     Monkey patching "multiprocessing" load environment.
Nuitka-Plugins:INFO: multiprocessing: Injecting post-module load code for module 'multiprocessing':
Nuitka-Plugins:INFO: multiprocessing:     Monkey patching "multiprocessing" for compiled methods.
Nuitka-Plugins:WARNING: Use '--plugin-enable=pylint-warnings' for: Understand PyLint/PyDev annotations for warnings.
Nuitka:INFO: Completed Python level compilation and optimization.
Nuitka:INFO: Generating C source code for backend compiler.
Nuitka:INFO: Running data composer tool for optimal constant value handling.
Nuitka:INFO: Running C level backend compilation via Scons.
Nuitka-Scons:INFO: Backend C compiler: cl (cl).
Traceback (most recent call last):

  File "c:\python37-64\lib\runpy.py", line 193, in _run_module_as_main

    "__main__", mod_spec)

  File "c:\python37-64\lib\runpy.py", line 85, in _run_code

    exec(code, run_globals)

  File "C:\Python37-64\scripts\clcache.exe\__main__.py", line 7, in <module>

  File "c:\python37-64\lib\site-packages\clcache\__main__.py", line 1577, in main

    return processCompileRequest(cache, compiler, sys.argv)

  File "c:\python37-64\lib\site-packages\clcache\__main__.py", line 1604, in processCompileRequest

    return scheduleJobs(cache, compiler, cmdLine, environment, sourceFiles, objectFiles)

  File "c:\python37-64\lib\site-packages\clcache\__main__.py", line 1656, in scheduleJobs

    exitCode, out, err, doCleanup = future.result()

  File "c:\python37-64\lib\concurrent\futures\_base.py", line 428, in result

    return self.__get_result()

  File "c:\python37-64\lib\concurrent\futures\_base.py", line 384, in __get_result

    raise self._exception

  File "c:\python37-64\lib\concurrent\futures\thread.py", line 57, in run

    result = self.fn(*self.args, **self.kwargs)

  File "c:\python37-64\lib\site-packages\clcache\__main__.py", line 1677, in processSingleSource

    return processDirect(cache, objectFile, compiler, cmdLine, sourceFile)

  File "c:\python37-64\lib\site-packages\clcache\__main__.py", line 1707, in processDirect

    return processCacheHit(cache, objectFile, cachekey)

  File "c:\python37-64\lib\site-packages\clcache\__main__.py", line 1495, in processCacheHit

    stats.registerCacheHit()

  File "c:\python37-64\lib\site-packages\clcache\__main__.py", line 753, in __exit__

    self._stats.save()

  File "c:\python37-64\lib\site-packages\clcache\__main__.py", line 663, in save

    json.dump(self._dict, f, sort_keys=True, indent=4)

  File "c:\python37-64\lib\contextlib.py", line 119, in __exit__

    next(self.gen)

  File "c:\python37-64\lib\site-packages\clcache\__main__.py", line 124, in atomicWrite

    os.replace(tempFileName, fileName)

PermissionError: [WinError 5] Acc\xe8s refus\xe9: 'C:\\Users\\Me\\clcache\\stats.txt.new' -> 'C:\\Users\\Me\\clcache\\stats.txt'
```

**Comment by deajan** [comment]
As for the clcache initial issue about the permission error, I could around by doing the following

```
SET PATH=%PATH%;C:\Users\myuser\AppData\Local\GitHubDesktop\app-2.6.1\resources\app\git\mingw64\bin
c:\Python38-64\python.exe -m pip install -U git+https://github.com/frerich/clcache.git#egg=proj
```

Since I'm working on Windows, I had to add the path to my `git.exe`
If we choose this as a current workaround, shall I make a doc PR ?

**Comment by deajan** [comment]
Well nope, still got a permission error from time to time on one, one one of my build stations with 16 cores where race conditions are more likely. Uninstalling clcache for now.

**Comment by kayhayen** [comment]
You can force it to use one currently with `NUITKA_CLCACHE_BINARY` specified. I have had issues with too old clache btw, specifically the one in Anaconda is too old, use the one from pip in any case.

I wouldn't use git to install, we can apply the same method as with Nuitka factory, can we not, there we do this:

pip install -U "https://github.com/Nuitka/Nuitka/archive/factory.zip"

Arguably, that is probably the test reaction, e.g. nowadays ccache is downloaded automatically too. Us executing that command would be no problem, except that the scons might not be running inside that virtualenv, and use the wrong Python for that, and it may not even be a supported version, I think 3.6 or higher should be used. 

Can you @deajan report which clcache versions these were? I am using only the latest and old stable. But I have had permission errors often, probably due to concurrent compiles, in fact so often, that the test runner checks for these and retries. The `clcache` is also unmantained, is it not? Last i checked it was archived on Github and that e.g. also might cause us to loose it eventuallyand when 3.9 will e.g. break it, how to tell.

Moving this extra ball stuff lower into say scons itself may also be an option.

Including clcache inside of Nuitka is also an option, I have age old permission from the author to do it, and that may in fact the easiest way. We could then also compile it outselves in accelerated mode for that Python when needed, currently I am using a pyinstaller accelerated binary from PyPI wheel, and that feels wrong for Nuitka. But we can definitely start out with any compilation at all, and use the Python files directly.

Essentially we could also try and fix the permission error in an inline copy of clcache, because how hard can it be?

Maybe you could start by investigating how different current git and last stable release are. I have not had luck contacting the author about it or finding any statements on it.

**Comment by kayhayen** [comment]
The resolution order currently is `PATH` search, then these:

```
def _getPythonDirCandidates(python_prefix):
    result = [python_prefix]

    for python_dir in (
        sys.prefix,
        os.environ.get("CONDA_PREFIX"),
        os.environ.get("CONDA"),
    ):
        if python_dir and python_dir not in result:
            result.append(python_dir)

    return result
```

That is the Python version used to run Nuitka, the one used to run Scons, then Anaconda environments. Same is done for `ccache.exe` btw, this code is reused. In the later case, I am also tempted to *only* accept the downloaded version, since I have seen `ccache` binaries that hang sometimes forever, e.g. ones from MSYS2, where that had been the one I had used before, but after an update, it wasn't good anymore.

**Comment by kayhayen** [comment]
I had originally considered to iterate registry and find all Python installations btw, so that any `clcache` installed anywhere could be detected.

**Comment by kayhayen** [comment]
I just had a realization. What we need to do here, is simple. We create a fork of clcache on Github in the Nuitka organisation, fix it up, e.g. being unable to provide a statistics file update, we do not have to care. If caching fails, then it does.

From that then we provide a wheel to install, changing it from pyinstaller to Nuitka and/or provide a `.dist` zip file to download with the usual means, like gcc and ccache and dependency walker are installed as well.

That should enable us to solve this very much perfectly I guess. What do you think @deajan ?

**Comment by deajan** [comment]
I used clcache v4.2.0 (the one that comes with pip) and the latest git release.
Indeed, forking and maintaining as 'mature' project would be a good idea IMHO.

**Comment by kayhayen** [comment]
I offered the take over, but a fork could also make sense, I can also imagine running the clcache code directly inside of scons, and then we could just make it special cases, e.g. we only update the cache when everything is done in one go, with only one lock take, and that would essentially replace the daemon use case. 

We are more allowing about errors, and so on. Let's see how the owner responds to my request. I also want to have a look at how pyinstaller is integrated there, and if maybe there is a way to simply hook this out with another target, or be compatible in how that is done.

**Comment by kayhayen** [comment]
Ok, this has been accepted, and I am  now going to follow the plan of having an inline copy for the Nuitka Windows release, and make it so that the Scons runs it *not* as an external binary, but inside the own process as a thread probably (things are already a thread there).

Then we could make it so, that one Nuitka compilation does at least not compete with itself over a lock file to update the statistics file that we do not care about as much, maybe even make it possible to disable that if not already the case.

**Comment by kayhayen** [comment]
There we go https://github.com/Nuitka/clcache

I deleted stale branches, and added a develop branch, where we should do our changes and then eventually release clcache with a version 5.

**Comment by kayhayen** [comment]
While making initial changes after takeover, e.g. updating URLs, I now noticed why an inline copy wasn't as good. I think clcache is not running on older versions, but I am not sure of that.

**Comment by kayhayen** [comment]
While making initial changes after takeover, e.g. updating URLs, I now noticed why an inline copy wasn't as good. I think clcache is not running on older versions, but I am not sure of that. We might explore the 3to2 route that I am considering for other Nuitka plugins too.

**Comment by kayhayen** [comment]
I have had success with making changes, so that when setting an environment variable, my Python3.8 would run `clcache` after importing to to the scons instance, and wow, was it fast in the cached case.

The plan is now to make this through an inline copy of clcache needed files, and to use that for those Python3 versions that can use it. Going back with Python versions with applying selective 3to2 fixers, although I imagine that e.g. removing annotations from the inline copy (it's not for reading) would e.g. be a good thing anyway.

With this, the lock file behavior for Nuitka may exponentially better, because it is only one program updating the cache typically instead of many instead of "your core many count ones" competing for the files. 

Using the server wasn't really an option, as there was no way of knowing when to stop it.

**Comment by kayhayen** [comment]
I pushed to factory the clcache inline copy. When using Python 3.6 or higher it's used unless overriden by `NUITKA_CLCACHE_BINARY`, but that should go away.

I have used 3to2 to trim the code of annotations, which only waste memory, and to apply some portability improvements, which I believe might suffice to make it work on Python2. The code was using a useless `yield from` with a generator expression, instead of just returning that, otherwise it's not really Python3 code at all I think, and we only use a small portion of it.

**Comment by kayhayen** [comment]
Ok, factory ought to be good now. I got clcache to work with Python2 easily, with just one more small change needed.

But the execution was flawed, in that our entry point didn't pass environment variables on, which needed clcache changes to pass it on to where the execution happens.

Also Python2.6 on Windows now needs a higher version Python installed for Scons, that is probably for the better, as it makes sure we only maintain one of the scons inline copies, the one for 2.7 or higher. I had previously kept the old one that was best patched and less buggy, but that has likely reversed due to bitrot. We do that for 3.3 and 3.4 already on all platforms, so that is not new.

The clcache stats file and the cache files have moved to appdata folder. The stats file is now per Nuitka build, and therefore should be impossible to have a lock collision as you had.

Once this proves to be working in my private CI, the old clcache searching code will be removed, and providing an external one will not be supported anymore. That definitely will make it less fuzzy and also less complex in terms of documentation and CI needs.

I have yet to try it with `clangcl.exe` which had changes on the branch of old clcache repo that were unreleased, but I had thought it was already working, but maybe not.

Also also many changes I made to clcache should have broken it for other purposes, but we need to get the tests up and running eventually. For now, I hope it's not too bad.

**Comment by kayhayen** [comment]
The the current pre-release has this included, and doesn't search for clcache outside of Nuitka anymore. The `clangcl.exe` works just fine with it. 

The code of clcache in Nuitka is 2.7 ready, but the locking doesn't work, due to different ways how threading works in Python2 and Python3. In the later, a system mutex acquired by a thread is not held by the process, in Python2 it is, breaking the locking between them. Since I don't really care, using a Python3 for scons is now mandatory and that's it. So 2.7 won't cut it there, but having a Python3 installed is no big requirement bump, is it, esp. since it's located automatically.

Glad how this turned out, clcache is now really fast to use with Nuitka, before it was very prone to hanging and locking issues, that I have not yet again yet.

**Comment by kayhayen** [comment]
Released as part of 0.6.11
